test/functional/tool_utils.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+# Copyright 2014 BitPay Inc.
+# Copyright 2016-present The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit.
+"""Exercise the utils via json-defined tests."""
+
+from test_framework.test_framework import BitcoinTestFramework
+
+import difflib
+import json
+import logging
+import os
+import subprocess
+from pathlib import Path
+
+
+class ToolUtils(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 0  # No node/datadir needed
+
+    def setup_network(self):
+        pass
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_bitcoin_tx()
+        self.skip_if_no_bitcoin_util()
+
+    def run_test(self):
+        self.testcase_dir = Path(self.config["environment"]["SRCDIR"]) / "test" / "util" / "data"
+        self.bins = self.get_binaries()
+        with open(self.testcase_dir / "bitcoin-util-test.json", encoding="utf8") as f:
+            input_data = json.loads(f.read())
+
+        for i, test_obj in enumerate(input_data):
+            self.log.debug(f"Running [{i}]: " + test_obj["description"])
+            self.test_one(test_obj)
+
+    def test_one(self, testObj):
+        """Runs a single test, comparing output and RC to expected output and RC.
+
+        Raises an error if input can't be read, executable fails, or output/RC
+        are not as expected. Error is caught by bctester() and reported.
+        """
+        # Get the exec names and arguments
+        if testObj["exec"] == "./bitcoin-util":
+            execrun = self.bins.util_argv() + testObj["args"]
+        elif testObj["exec"] == "./bitcoin-tx":
+            execrun = self.bins.tx_argv() + testObj["args"]
+
+        # Read the input data (if there is any)
+        inputData = None
+        if "input" in testObj:
+            with open(self.testcase_dir / testObj["input"], encoding="utf8") as f:
+                inputData = f.read()
+
+        # Read the expected output data (if there is any)
+        outputFn = None
+        outputData = None
+        outputType = None
+        if "output_cmp" in testObj:
+            outputFn = testObj['output_cmp']
+            outputType = os.path.splitext(outputFn)[1][1:]  # output type from file extension (determines how to compare)
+            try:
+                with open(self.testcase_dir / outputFn, encoding="utf8") as f:
+                    outputData = f.read()
+            except Exception:
+                logging.error("Output file " + outputFn + " cannot be opened")
+                raise
+            if not outputData:
+                logging.error("Output data missing for " + outputFn)
+                raise Exception
```

Comment:
In faa18bf287fc02339b7af29d54af862a289bf96d "test: Turn util/test_runner into functional test"

This could be a good time to move the error message into the Exception rather than an error log.